bitwarden/browser
# [PS-102] BEEEP: Tab-change performance improvements

## Type of change


## Objective
As reported by some users the extension is quite active even while not visiting new sites as it reacts to all tab changes. This includes multiple state changes while a tab is loading but also the multiple state changes to tabs other than the currently focused tab (also across multiple windows).

This can be noticed by CPU spikes, higher memory usage and a higher amount of idle wake-up-calls.

This PR aims to reduce the calls to `refreshBadgeAndMenu` which seems to have the highest impact on the above mentioned values. Also not checking the `notificationQueue` for new logins tabs that currently have no interaction by the user and the collection of pageDetails for auto-fill.

CPU spikes and memory consumption are expected when the extension is performing operations in the background as for example: `sync`, `checking vault timeout`, `scanning a page for changes/login fields` but with the reduction in calls of the latter, this should improve the overall performance.

Closes #769 and #2198 

I hope this will also solve cases where the badge and menu do not correspond to the vault's lock state (unlocked/locked):
https://github.com/bitwarden/browser/issues/1717
https://github.com/bitwarden/browser/issues/2323
https://github.com/bitwarden/browser/issues/2455

Asana task: https://app.asana.com/0/1200578627114106/1201993816674939/f

## Code changes
- **src/background/main.background.ts:** 
  - Do not pass in `this` (main) to `notification.background`
  - Remove instantiation of `windows.background` - logic moved into `tabs.background`
  - Pass `true` to `refreshBadgeAndMenu` on `logout` which shortens the time to re-draw badge and menu
- **src/background/notification.background.ts:**
  - Remove constructor dependency on `main.brackground`
- **src/background/tabs.background.ts:** 
  - Listen to `chrome.windows.onFocusChanged` to set currently focusedWindowId
  - Remove messages send that nothing subscribes to
  - Only react to tab changes of currently focused window
  - Check status of tab and onlt react to change if tab is active
- **src/background/windows.background.ts:** 
  - Deleted, as logic has been moved to `tabs.background`\
- **package-lock.json:** Changes after running `npm i`


## Screenshots
I've monitored the changes using the built-in task manager in Chrome/Firefox and below are comparision videos for before and after my applied changes.

As a testbed, I always used a freshly started chrome/firefox profile, installing the dev-extension, closing the welcome page and opening 10 tabs of [https:bigstar](https://bigstar.pl/) (as this site changes it's title when not in focus) and one focused new tab (browser default page).


### Firefox
[Before video](https://user-images.githubusercontent.com/2670567/159126792-8581a50d-1edd-49b5-955b-dce2c42b4050.mp4)
[After video](https://user-images.githubusercontent.com/2670567/159126804-c70e1757-f1df-4c34-b285-53a13c956b89.mp4)

### Chrome
[Before video](https://user-images.githubusercontent.com/2670567/159126671-4f13ca24-df58-49c2-93a3-45e22c568aca.mp4)
[After video](https://user-images.githubusercontent.com/2670567/159126663-b4a7f42f-b2aa-49ac-9925-765244922a84.mp4)

### Safari
As I am currently unable to test my changes on Mac/Safari, I'd love for one of the devs to test this, before progressing to QA.

## Testing requirements
Testing should occur with all of our supported browsers, and include multiple open windows and tabs and jumping between them while checking the test cases below:
- The badge and menu should always show the correct state, between un-/authed, un-/locked
- The badge should always show the correct amount of matched entries
- Make sure auto-fill still works as expected 
- Make sure unlock on autofill works, shows the correct states on login and uses the possibly last auto-filled entry
- Changes to tabs in the same window should not create cpu-spikes (easiest to monitor)
- Changes to tabs in a different window should not create cpu-spikes (easiest to monitor)

If any further questions arise, feel free to reach out.

## Before you submit
